cpp/ql/lib/semmle/code/cpp/models/interfaces/FormattingFunction.qll

```diff
@@ -121,16 +121,31 @@ abstract class FormattingFunction extends ArrayFunction, TaintFunction {
    * the first format specifier in the format string.
    */
   int getFirstFormatArgumentIndex() {
-    result = this.getNumberOfParameters() and
-    // the formatting function either has a definition in the snapshot, or all
+    // The formatting function either has a definition in the snapshot, or all
     // `DeclarationEntry`s agree on the number of parameters (otherwise we don't
     // really know the correct number)
-    (
-      this.hasDefinition()
-      or
-      forall(FunctionDeclarationEntry fde | fde = this.getADeclarationEntry() |
-        result = fde.getNumberOfParameters()
-      )
+    if this.hasDefinition()
+    then result = this.getDefinition().getNumberOfParameters()
+    else result = this.getNumberOfExplicitParameters()
+  }
+
+  /**
+   * Gets a non-implicit function declaration entry.
+   */
+  FunctionDeclarationEntry getAnExplicitDeclarationEntry() {
+    result = this.getADeclarationEntry() and
+    not result.isImplicit()
+  }
```

Comment:
This feels like it could be a method of `Function`, rather than `FormattingFunction`?

Reply:
Shouldn't this just be made `private` (same for the other predicate)?

Reply:
I did slightly wonder if this would be useful elsewhere, particularly because the number of implicit function declarations is likely to increase significantly for BMN. But I wasn't quite brave enough to put it on `Function` without more evidence it would be useful.

Reply:
Keep them here, make them private (so we're not changing any interface yet), and move them once we see we need them in other places?